mllib/src/main/scala/org/apache/spark/ml/tree/Split.scala

```diff
@@ -145,6 +147,11 @@ final class CategoricalSplit private[ml] (
   private def setComplement(cats: Set[Double]): Set[Double] = {
     Range(0, numCategories).map(_.toDouble).filter(cat => !cats.contains(cat)).toSet
   }
+
+  override def hashCode(): Int = {
+    val state = Seq(isLeft, categories, featureIndex, numCategories)
```

Comment:
This could go near `equals`. You're using an extra field not used in equals. It may work out OK but maybe easier to make them match.